src/lib/updating-element.ts

```diff
@@ -0,0 +1,503 @@
+ */
+
+/**
+ * Converts property values to and from attribute values.
+ */
+type AttributeProcessor = {
+  /**
+   * Deserializing function called to convert an attribute value to a property value.
+   */
+  fromAttribute?(value: string): unknown,
+  /**
+   * Serializing function called to convert a property value to an attribute value.
+   */
+  toAttribute?(value: unknown): string|null
+};
+
+/**
+ * Defines options for a property accessor.
+ */
+export interface PropertyOptions {
+  /**
+   * Describes how and if the property should becomes an observedAttribute.
+   * If explicitly false, the property will not be observed. Otherwise if true
+   * or absent, the lowercased property name is observed, and if a string
+   * observe that value.
+   */
+  attribute?: boolean|string;
+  /**
+   * Describes how to convert a property value to and from an attribute. If
+   * the value is a function, it's calle to deserialize an attribute value into
+   * a property value. If it's an AttributeProcessor, the `fromAttribute` value
+   * is used to deserialize. If the `reflect` option is also true, then
+   * `toAttribute` value is used to serialize the property value to an attribute.
+   */
+  type?: AttributeProcessor|Function;
+  /**
+   * Describes if setting the property should be reflect to the attribute value.
+   * If true, then if present the `type.toAttribute` function is used to serialize
+   * the value; otherwise, the property value is used.
+   */
+  reflect?: boolean;
+  /**
+   * Describes if setting a property should trigger invalidation and updating.
+   * This function takes the new and oldValue and returns true if invalidation
+   * should occur. If not present, a strict identity check is used.
+   */
+  shouldInvalidate?(value: unknown, oldValue: unknown): boolean;
+}
+
+/**
+ * Object that describes accessors to be created on the element prototype.
+ * An accessor is created for each key with the given options.
+ */
+export interface Properties {
+  [key: string]: PropertyOptions;
+}
+
+interface AttributeMap {
+  [key: string]: string;
+}
+
+interface AnyObject {
+  [key: string]: {};
+}
+
+/**
+ * Creates a property accessor on the given prototype if one does not exist.
+ * Uses `getProperty` and `setProperty` to manage the property's value.
+ */
+const makeProperty = (name: string, proto: Object) => {
+  if (proto.hasOwnProperty(name) || (name in proto)) {
+    return;
+  }
+  Object.defineProperty(proto, name, {
+    get() {
+      return this.getProperty(name);
+    },
+    set(value) {
+      this.setProperty(name, value);
+    },
+    configurable: true,
+    enumerable: true
+  });
+};
+
+/**
+ * Creates and sets object used to memoize all class property values. Object
+ * is chained from superclass.
+ */
+const ensurePropertyStorage = (ctor: typeof UpdatingElement) => {
+  if (!ctor.hasOwnProperty('_classProperties')) {
+    ctor._classProperties = Object.create(Object.getPrototypeOf(ctor)._classProperties);
```

Comment:
Don't you need to use `defineProperty` here in the case that the super-class already has a `_classProperties` property? Do the tests cover subclasses and ensuring that subclass properties don't get defined on superclasses?

Reply:
Setting the property should shadow any superclass, but will add a test.